packages/jest-runtime/src/index.ts

```diff
@@ -523,7 +502,9 @@ export default class Runtime {
 
     const modulePath = this._resolveModule(from, path);
 
-    return this.loadEsmModule(modulePath, query);
+    const module = await this.loadEsmModule(modulePath, query);
+
+    return this.linkAndEvaluateModule(module);
```

Comment:
bug fix is mostly here - only call `linkAndEvaluateModule` on the entry point, plus dynamic imports. The old `loadEsmModule` now just creates the module and sticks it in the cache